expand cluster without internet access test and move slurm specific dna.json attributes to a separate function

hanwen-pcluste opened the pull request:
### Description of changes
* Describe *what* you're changing and *why* you're doing these changes.

The workaround is in Cookbook. This commit will expand the `test_cluster_in_no_internet_subnet` test to verify `srun` with `OpenMPI`, which the workaround will be able to address.

This commit will also move Slurm specific dna.json attributes to a separate function, aiming at simplifying the code

### Tests
* Describe the automated and/or manual tests executed to validate the patch.
`test_cluster_in_no_internet_subnet` is the integration tests for this feature.
I also manually submitted 150 OSU benchmark jobs that require 76 nodes in total. The jobs succeeded

### References
* Link to related PRs in other packages (i.e. cookbook, node).
This PR only works with https://github.com/aws/aws-parallelcluster-cookbook/pull/1310

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

demartinofra commented on cli/src/pcluster/templates/cdk_builder_utils.py:
```diff
@@ -109,6 +110,17 @@ def get_common_user_data_env(node: Union[HeadNode, SlurmQueue], config: BaseClus
     }
 
 
+def get_slurm_specific_dna_json_for_head_node(config: SlurmClusterConfig, scheduler_resources) -> dict:
```
nit: shall we move this to cluster_stack.py since only used in there?

hanwen-pcluste replied:
We somehow have been adding utils functions in this file to keep cluster_stack.py clean

demartinofra replied:
I'm ok with leaving it here but in general when the code is used in a single place and is very specific the close we keep its definition to where it's used the better.

demartinofra commented on cli/src/pcluster/resources/compute_node/user_data.sh:
```diff
@@ -87,6 +87,7 @@ write_files:
           "custom_node_package": "${CustomNodePackage}",
           "custom_awsbatchcli_package": "${CustomAwsBatchCliPackage}",
           "use_private_hostname": "${UsePrivateHostname}",
+          "workaround_with_private_hostname": "${WorkaroundWithPrivateHostname}",
```
Can we keep this variable only in cookbook without explicitly setting the default in the user data and relying on use_private_hostname in cookbook for the default? Also I'd rename it to something like: `enable_node_hostnames_local_resolution`.

hanwen-pcluste replied:
Done. The CLI no long explicitly set this variable

hanwen-cluster committed:
integ-test: expand cluster without internet access test

The workaround is in Cookbook. This commit will expand the `test_cluster_in_no_internet_subnet` test to verify `srun` with `OpenMPI`, which the workaround will be able to address.

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Move Slurm specific dna.json attributes to a separate function for simplification

Signed-off-by: Hanwen <hanwenli@amazon.com>

demartinofra commented on tests/integration-tests/tests/networking/test_cluster_networking/test_cluster_in_no_internet_subnet/osu_pt2pt_submit_openmpi.sh:
```diff
@@ -4,4 +4,5 @@
 set -e
 
 module load openmpi
+srun --mpi=pmix /shared/openmpi/osu-micro-benchmarks-{{ osu_benchmark_version }}/mpi/pt2pt/osu_latency
```
optional: leaving the comment here but it's about the test more in general. Ideally it would be nice to verify that when the job completes the /etc/hosts file is properly cleaned up. Even better if we could submit 2 jobs and check that the entries are properly prepended.

hanwen-pcluste replied:
Thank you! I would address this in a separate PR

enrico-usai commented on cli/src/pcluster/templates/cdk_builder_utils.py:
```diff
@@ -109,6 +110,16 @@ def get_common_user_data_env(node: Union[HeadNode, SlurmQueue], config: BaseClus
     }
 
 
+def get_slurm_specific_dna_json_for_head_node(config: SlurmClusterConfig, scheduler_resources) -> dict:
```
minor: what about `get_slurm_dna_json_for_head_node`

enrico-usai committed:
Merge branch 'develop' into nointernetaccesspure
